### PR TITLE
[14.0] [IMP] sale_order_product_assortment: Use original product_id field keeping domain. Use cache to performance

### PR DESCRIPTION
**Pull request by renda-dev**
Back port of https://github.com/OCA/sale-workflow/pull/2538

Depends on:

- [x] https://github.com/OCA/server-tools/pull/2941
- [x] https://github.com/OCA/server-tools/pull/2942

Edit: I've also added a commit that changes the AND in the `_compute_product_assortment_ids` into an OR since, in my opinion, it's more likely what an user would expect by combining more assortments

### CODE REVIEW

**Review by francesco-ooops — APPROVED** [review]
Functional ok, good for merge once depending PR is merged

**Comment by francesco-ooops** [comment]
@CarlosRoca13 what do you think?

We had to apply the latest commit as the behavior was inconsistent, eg: 

Create assortment "Chair", product allowed: "Office Chair", partner: Azure Interior
Create assortment "Screen", product allowed: "Acoustic bloc screens", partner: Azure Interior
SO for Azure Interior: 2 products can be selected

Create assortment "Chair", allowed product domain: "Name contains Office Chair", partner: Azure Interior
Create assortment "Screen", allowed product domain: "Name contains Acoustic bloc screens", partner: Azure Interior
SO for Azure Interior: no products can be selected

We'll fw-port this commit

**Review by HekkiMelody — APPROVED** [review]
Code review, LGTM.

I think it makes more sense and is more intuitive for the user to have different assortments linked with OR than with AND. Otherwise the effect is that, as you add more assortments to the customer, you actually shrink the selection they have available.

**Comment by CarlosRoca13** [comment]
Makes sense. Can you please add in the commit a brief explanation and the steps to reproduce the problem?

**Review by CarlosRoca13 — COMMENTED** [review]
On second thought... Let's imagine that we have 2 assortments for the same customer, in these two assortments, what is done, is to ban the purchase of chairs in the first one and tables in the second one. With this change, the customer will be able to see all the products, both chairs and tables.

**Comment by francesco-ooops** [comment]
@CarlosRoca13 yours is a valid point, but it seems like this is a short blanket situation, as some assortments will be used to exclude products and others to include them and some to do both - there is no right or wrong. 

We will look into a solution for the matter, do you have any idea how this could be solved?

**Comment by CarlosRoca13** [comment]
Right now, I don't know how this could be faced... But I know that we have customers who are using the assortments excluding products and we cannot assume this change...

@pedrobaeza @carlosdauden @sergio-teruel can you give your point of view about this?

**Comment by pedrobaeza** [comment]
There shouldn't be any feature regression. It's OK to optimize, but not losing features.

**Comment by francesco-ooops** [comment]
> There shouldn't be any feature regression. It's OK to optimize, but not losing features.

We're on the same page on that (we've always been), but we need to address the inconsistency in the example I posted above:

> Create assortment "Chair", product allowed: "Office Chair", partner: Azure Interior
> Create assortment "Screen", product allowed: "Acoustic bloc screens", partner: Azure Interior
> SO for Azure Interior: 2 products can be selected
> 
> Create assortment "Chair", allowed product **domain**: "Name contains Office Chair", partner: Azure Interior
> Create assortment "Screen", allowed product **domain**: "Name contains Acoustic bloc screens", partner: Azure Interior
> SO for Azure Interior: no products can be selected

Also note that website assortment module works with OR as per https://github.com/OCA/e-commerce/pull/838

**Comment by pedrobaeza** [comment]
Are you telling that previously it was not correct? I agree then to cover the case correctly with the optimization.

**Comment by CarlosRoca13** [comment]
The real problem of this is that we are taking all assortment domain at this point: 
https://github.com/OCA/sale-workflow/pull/3135/commits/7e8b7a484f85ed4fb4be881b95ac8ea9dc4f921c#diff-d857107550de88517b0f4cb32ae7770a25a630dc254b97191e7f71f262183089R36
Since we don't know what type of assortment is involved

I think the right solution would be to take 2 product domains out of the assortment. One for excluded products and one for included products. In this way we could combine them in the way that suits us.

Note: The excluded products can't fit with the included with OR in the domain

**Comment by francesco-ooops** [comment]
Sounds like something we could think about!

> Note: The excluded products can't fit with the included with OR in the domain

@CarlosRoca13 Can you expand on this?

**Comment by CarlosRoca13** [comment]
4 sure, what I mean is that the excluded domain needs to be set with AND because it is restrictive, an excluded product discards any posible included product

**Comment by francesco-ooops** [comment]
@CarlosRoca13 @pedrobaeza 

Ok so the idea is to have in each assortment one "inclusive" domain working with OR and an "exclusive" domain working with AND, is that correct?

Maybe for user clarity we can specify below the domains the logic that will be applied, would that work?

**Comment by pedrobaeza** [comment]
At the end, both should combine to get a unique domain for the product selection, but include part should be treated on one part, joining all the inclusions together, and exclusions should be also joined, and excluded from the remaining result.

File: sale_order_product_assortment/models/sale_order.py [hunk 1]
```diff
@@ -16,27 +19,61 @@ class SaleOrder(models.Model):
...
-            if self.allowed_product_ids:
+                ]
+                product_domain = expression.AND(
+                    [whitelist_product_domain, blacklist_product_domain]
+                )
+            if product_domain:
+                self.allowed_product_ids = self.env["product.product"].search(
+                    product_domain
+                )
                 self.has_allowed_products = True
```
Comment (HekkiMelody) [inline sale_order_product_assortment/models/sale_order.py:76]: **chore**: `if self.allowed_product_ids` should probably be preserved here.

**Review by HekkiMelody — CHANGES_REQUESTED** [review]
Code review seems ok, I definitely prefer this solution (add all whitelisted products, remove all blacklisted products) but I think we should add a couple of tests here to make sure.

Also, outside of this PR, the logic should be unified and moved to the base module so that there are no differences with the e-commerce module.

**Review by HekkiMelody — APPROVED** [review]
Code review, LGTM

**Review by CarlosRoca13 — APPROVED** [review]
LGTM :+1: 

Please fw-port changes to 15.0 but note that a domain has been introduced for excluded products.

https://github.com/OCA/product-attribute/blob/50e3e48b2fcf5f504a857b367dd90e7ed7812fd0/product_assortment/models/ir_filters.py#L75

File: sale_order_product_assortment/models/sale_order.py [hunk 2]
```diff
@@ -28,17 +29,52 @@ def _compute_product_assortment_ids(self):
         product_domain = []
         if self[partner_field]:
             filters_partner_domain = self.env["ir.filters"].search(
-                [("is_assortment", "=", True)]
+                [
+                    ("is_assortment", "=", True),
```
Comment (pedrobaeza) [inline sale_order_product_assortment/models/sale_order.py]: Don't introduce a comma here for avoiding this unneeded diff.

Comment (renda-dev) [reply sale_order_product_assortment/models/sale_order.py]: Fixed, thanks

**Pull request merged**